service/history/engine/engineimpl/reset_workflow_execution_test.go

```diff
@@ -0,0 +1,499 @@
+						gomock.Any(), // Context
+						gomock.Eq(constants.TestDomainID),
+						gomock.Eq(constants.TestWorkflowID),
+						gomock.Eq(latestExecution.RunID),
+						gomock.Eq([]byte("branch token")),
+						gomock.Eq(int64(99)),   // Request.DecisionFinishEventID - 1
+						gomock.Eq(int64(1337)), // CurrentVersion
+						gomock.Eq(int64(100)),  // NextEventID
+						gomock.Any(),           // random uuid
+						gomock.Eq(testRequestID),
+						&workflowMatcher{latestExecution},
+						gomock.Eq(testRequestReason),
+						gomock.Nil(),
+						gomock.Eq(testRequestSkipSignalReapply),
+					).Return(nil).Times(1)
+				},
+			},
+			// Can't assert on the result because the runID is random
+		},
+		{
+			name: "Success using version histories",
+			// This corresponds to VersionHistories.Histories.Items.EventID
+			request: resetExecutionRequest(latestExecution, 50),
+			init: []InitFn{
+				withCurrentExecution(latestExecution),
+				withState(latestExecution, &persistence.WorkflowMutableState{
+					ExecutionInfo: &persistence.WorkflowExecutionInfo{
+						DomainID:    constants.TestDomainID,
+						WorkflowID:  constants.TestWorkflowID,
+						RunID:       latestRunID,
+						NextEventID: 100,
+						BranchToken: []byte("branch token"),
+					},
+					VersionHistories: &persistence.VersionHistories{
+						CurrentVersionHistoryIndex: 1,
+						Histories: []*persistence.VersionHistory{
+							{
+								BranchToken: []byte("this one isn't current"),
+							},
+							{
+								BranchToken: []byte("yes"),
+								Items: []*persistence.VersionHistoryItem{
+									{
+										EventID: 1,
+										Version: 1001,
+									},
+									{
+										EventID: 50,
+										Version: 1002,
+									},
+									{
+										EventID: 51,
+										Version: 1003,
+									},
+								},
+							},
+						},
+					},
+					ExecutionStats: &persistence.ExecutionStats{HistorySize: 1},
+				}),
+				func(t *testing.T, engine *testdata.EngineForTest) {
+					ctrl := gomock.NewController(t)
+					mockResetter := reset.NewMockWorkflowResetter(ctrl)
+					engine.Engine.(*historyEngineImpl).workflowResetter = mockResetter
+
+					mockResetter.EXPECT().ResetWorkflow(
+						gomock.Any(), // Context
+						gomock.Eq(constants.TestDomainID),
+						gomock.Eq(constants.TestWorkflowID),
+						gomock.Eq(latestExecution.RunID),
+						gomock.Eq([]byte("yes")), //VersionHistories.Histories.BranchToken
+						gomock.Eq(int64(49)),     // Request.DecisionFinishEventID - 1
+						gomock.Eq(int64(1002)),   // VersionHistories.Histories.Items.Version
+						gomock.Eq(int64(100)),    // NextEventID
+						gomock.Any(),             // random uuid
+						gomock.Eq(testRequestID),
+						&workflowMatcher{latestExecution},
+						gomock.Eq(testRequestReason),
+						gomock.Nil(),
+						gomock.Eq(testRequestSkipSignalReapply),
+					).Return(nil).Times(1)
+				},
+			},
+			// Can't assert on the result because the runID is random
```

Comment:
we could make `RunIDGenerator` a separate interface and mock it but doesn't worth it I guess.

Reply:
as a followup, it'd be nice to pull sources of entropy and side-effects into embedable inputs like the mockabe clock